### DISCUSSION TITLE
Why is Roslyn not automatically creating `IsExternalInit` if it doesn't find one?

### DISCUSSION THREAD

**Discussion by teo-tsirpanis** [dotnet/roslyn#61841] [Q&A] [answered]
If you define a `readonly` or `ref` `struct` on frameworks without `IsReadOnlyAttribute` or `IsByRefLikeAttribute`, Roslyn will happily compile it and emit internal attribute types in the assembly, completely transparently for us.

The same does not happen when defining `init` properties. Their setters are marked with a modreq of type `System.Runtime.CompilerServices.IsExternalInit`, which is available since .NET 5. If we are on an earlier framework however, Roslyn will complain that such attribute does not exist, requiring users to define an empty class with that name. This annoyance affects both first and third-party code (for example I have seen this class defined in source generators which are currently locked to .NET Standard 2.0).

I am wondering why doesn't Roslyn automatically generate an internal `IsExternalInit` class if it doesn't find one in either the core library or the compilation's source? Looking at the specs and past issues did not give me any answers.

**Comment by CyrusNajmabadi** [comment, marked as answer]
@333fred for info if I'm wrong. However, afaik the issue is this:

> If we are on an earlier framework however,

This is explicitly not a supported scenario. Which means any behavior is not defined and no intentional effort is made towards things working one way or the other. 

Even if this worked on some fashion, it might break at a moments notice. Ergo, any effort or attempt to have this work is *better* if explicitly done by the user so they recognize what they're getting themselves into.

**Reply by 333fred** [reply]
This is correct. Ref structs were introduced in C# 7, and were explicitly supported on older Framework versions. Thus, they had to work, and that means generating the attribute. Nullable too was initially designed this way (before we cut official Framework support) and the remnants of that are why the NullableAttribute is synthesized. However, new features (including init, required, etc) are not supported on older TFMs, and any support that can be shimmed in for those older TFMs is a happy accident, not an intentional design decision.

**Reply by teo-tsirpanis** [reply]
Thanks for the answers to both of you.
